File: lmfetch/analyzers/llm.py
"""LLM-powered smart analysis using fast/cheap models."""

import asyncio


async def summarize_chunk(content: str, max_length: int = 200) -> str:
    """Summarize a code chunk using a fast model."""
    try:
        from ai_query import generate_text, google

        result = await generate_text(
            model=google("gemini-2.0-flash-lite"),
            system="You are a code summarizer. Output ONLY a brief 1-2 sentence summary of what this code does. No markdown, no preamble.",
            prompt=f"Summarize this code:\n\n{content[:3000]}",
        )
        return result.text.strip()[:max_length]
    except Exception:
        # Fallback: first line or docstring
        lines = content.strip().split("\n")
        for line in lines[:5]:
            line = line.strip()
            if line.startswith('"""') or line.startswith("'''"):
                return line.strip("\"'")[:max_length]
            if line.startswith("#") or line.startswith("//"):
                return line.lstrip("#/ ")[:max_length]
        return lines[0][:max_length] if lines else ""


async def compute_relevance_score(query: str, content: str) -> float:
    """Use fast model to score relevance of content to query."""
    try:
        from ai_query import generate_text, google

        result = await generate_text(
            model=google("gemini-2.0-flash-lite"),
            system="You are a relevance scorer. Output ONLY a number from 0.0 to 1.0 indicating how relevant the code is to the query. Just the number, nothing else.",
            prompt=f"Query: {query}\n\nCode:\n{content[:2000]}",
        )
        score = float(result.text.strip())
        return max(0.0, min(1.0, score))
    except Exception:
        return 0.5


async def expand_query(query: str) -> list[str]:
    """Expand a query into related search terms."""
    try:
        from ai_query import generate_text, google

        result = await generate_text(
            model=google("gemini-2.0-flash-lite"),
            system="You expand search queries for code search. Given a query, output 3-5 related terms/phrases that would help find relevant code. Output one per line, no numbers or bullets.",
            prompt=f"Expand this code search query: {query}",
        )
        terms = [t.strip() for t in result.text.strip().split("\n") if t.strip()]
        return [query] + terms[:4]
    except Exception:
        return [query]


async def batch_summarize(chunks: list[tuple[str, str]], concurrency: int = 5) -> dict[str, str]:
    """Summarize multiple chunks concurrently."""
    semaphore = asyncio.Semaphore(concurrency)

    async def summarize_with_limit(path: str, content: str) -> tuple[str, str]:
        async with semaphore:
            summary = await summarize_chunk(content)
            return path, summary

    results = await asyncio.gather(*[
        summarize_with_limit(path, content)
        for path, content in chunks
    ])

    return dict(results)


async def rerank_with_llm(
    query: str,
    chunks: list[tuple[str, str, float]],  # (path, content, initial_score)
    top_k: int = 20,
) -> list[tuple[str, str, float]]:
    """Rerank top chunks using LLM for better relevance."""
    # Only rerank top candidates
    candidates = sorted(chunks, key=lambda x: x[2], reverse=True)[:top_k * 2]

    async def score_chunk(path: str, content: str, initial: float) -> tuple[str, str, float]:
        llm_score = await compute_relevance_score(query, content)
        # Blend initial score with LLM score
        final = initial * 0.4 + llm_score * 0.6
        return path, content, final

    reranked = await asyncio.gather(*[
        score_chunk(path, content, score)
        for path, content, score in candidates
    ])

    return sorted(reranked, key=lambda x: x[2], reverse=True)

File: lmfetch-bun/src/analyzers/llm.ts
/**
 * LLM-powered reranking (optional)
 */
import { generateText } from "ai";
import { google } from "@ai-sdk/google";
import { retry } from "../utils";
import type { ScoredChunk } from "../chunkers/types";

const RERANK_MODEL = "gemini-2.5-flash-lite";
const MAX_CHUNKS_TO_RERANK = 20;

/**
 * Use LLM to rerank chunks for better relevance
 */
export async function llmRerank(
  chunks: ScoredChunk[],
  query: string,
  onProgress?: (message: string) => void
): Promise<ScoredChunk[]> {
  const progress = onProgress || (() => {});

  // Only rerank top chunks
  const toRerank = chunks.slice(0, MAX_CHUNKS_TO_RERANK);
  const rest = chunks.slice(MAX_CHUNKS_TO_RERANK);

  progress("Running LLM reranking...");

  // Score each chunk with LLM
  const rerankedScores = await Promise.all(
    toRerank.map(async (chunk) => {
      const score = await scoreWithLLM(chunk, query);
      return { ...chunk, score };
    })
  );

  // Sort reranked chunks by new score
  rerankedScores.sort((a, b) => b.score - a.score);

  // Combine with rest (which keep their original relative order)
  return [...rerankedScores, ...rest];
}

/**
 * Score a single chunk using LLM
 */
async function scoreWithLLM(chunk: ScoredChunk, query: string): Promise<number> {
  try {
    const { text } = await retry(
      async () => {
        return await generateText({
          model: google(RERANK_MODEL),
          system: `You are a relevance scorer. Given a code snippet and a query, output ONLY a number from 0.0 to 1.0 indicating how relevant the code is to answering the query.

0.0 = completely irrelevant
0.5 = somewhat relevant
1.0 = highly relevant

Output ONLY the number, nothing else.`,
          prompt: `Query: ${query}

Code (${chunk.relativePath}):
\`\`\`${chunk.language}
${chunk.content.slice(0, 2000)}
\`\`\`

Relevance score:`,
          maxTokens: 10,
        });
      },
      { retries: 2, delay: 500 }
    );

    const score = parseFloat(text.trim());
    if (isNaN(score) || score < 0 || score > 1) {
      return chunk.score; // Keep original score if parsing fails
    }

    // Blend LLM score with original score
    return score * 0.7 + chunk.score * 0.3;
  } catch {
    return chunk.score; // Keep original score on error
  }
}
